### PR TITLE
rpc: faster getblockstats using BlockUndo data

### PR DESCRIPTION
**Pull request by FelixWeis**
Using undo data for a block (rev?????.dat) we can retrieve value information about prevouts and calculate the final transaction fee (rate). This approach is about 80x faster, drops the requirement for `-txindex`, and works for all non-pruned blocks.

```
# 2018-11-25T16:36:19Z Bitcoin Core version v0.17.99.0-edc715240-dirty (release build)
seq 550100 550200  0.00s user 0.00s system 62% cpu 0.004 total
xargs -n1 src/bitcoin-cli getblockstats  0.21s user 0.19s system 17% cpu 2.302 total


# 2018-11-25T16:39:17Z Bitcoin Core version v0.17.0 (release build)
seq 550100 550200  0.00s user 0.00s system 87% cpu 0.002 total
xargs -n1 src/bitcoin-cli getblockstats  0.24s user 0.22s system 0% cpu 3:19.42 total
```

### CODE REVIEW

**Comment by gmaxwell** [comment]
<3  I like this direction, certainly having a RPC that requires txindex is unfortunate, as does having rpcs that don't work (well) with pruning.  I don't think it would have been worth it to store extra data to make stats fast, but since we already are...

**Review by maflcko — COMMENTED** [review]
Concept ACK

File: src/validation.cpp [hunk 1]
```diff
@@ -1451,6 +1451,36 @@ bool CheckInputs(const CTransaction& tx, CValidationState &state, const CCoinsVi
     return true;
 }
 
+bool UndoReadFromDisk(CBlockUndo& blockundo, const CBlockIndex* pindex)
```
Comment (promag) [inline src/validation.cpp]: Could split move only to a different commit.

File: src/validation.h [hunk 2]
```diff
@@ -388,12 +389,15 @@ class CScriptCheck
 void InitScriptExecutionCache();
 
 
+
```
Comment (promag) [inline src/validation.h]: Remove change.

**Review by promag — COMMENTED** [review]
Great change!

Mind adding a release note?

**Comment by jonasschnelli** [comment]
Clever idea!
Concept ACK

**Comment by ajtowns** [comment]
ConceptACK, code looks good at first glance

**Comment by laanwj** [comment]
Good to remove reliance on the transaction index.
utACK 602dd542a80d76225285ad9d52ebfc2a4d4e7c54

File: test/functional/rpc_getblockstats.py [hunk 3]
```diff
@@ -128,8 +113,8 @@ def run_test(self):
             assert_equal(stats_by_hash, self.expected_stats[i])
 
             # Check with the node that has no txindex
```
[the remark this answers is not in the document]
Comment (FelixWeis) [reply test/functional/rpc_getblockstats.py]: test and testdata changed

File: src/validation.cpp (hunk 1 above)
Comment (FelixWeis) [reply src/validation.cpp]: it will get squashed anyway no?

File: src/validation.h [hunk 4]
```diff
@@ -19,6 +19,7 @@
 #include <script/script_error.h>
 #include <sync.h>
 #include <versionbits.h>
+#include <undo.h>
```
Comment (MarcoFalke) [inline src/validation.h]: style-nit: Could use a forward decl instead?

File: doc/release-notes.md [hunk 5]
```diff
@@ -298,6 +298,9 @@ RPC
 - A new `submitheader` RPC allows submitting block headers independently
   from their block.  This is likely only useful for testing.
 
+- `getblockstats` speedup and drops `-txindex` requirement, now works on
+  all non-pruned blocks.
+
```
Comment (marcinja) [inline doc/release-notes.md]: nit: This should be a full sentence.

**Review by marcinja — COMMENTED** [review]
utACK 76fef2639e

Nice work!

File: doc/release-notes.md (hunk 5 above)
Comment (MarcoFalke) [reply doc/release-notes.md]: And a separate file `./doc/release-notes-14802.md`

File: src/rpc/blockchain.cpp [hunk 6]
```diff
@@ -1987,6 +1982,7 @@ static UniValue getblockstats(const JSONRPCRequest& request)
             maxfeerate = std::max(maxfeerate, feerate);
             minfeerate = std::min(minfeerate, feerate);
         }
+        tx_n++;
```
Comment (MarcoFalke) [inline src/rpc/blockchain.cpp]: looks fragile. (If some paths in the for loop `continue`, this would be skipped)

File: src/rpc/blockchain.cpp [hunk 7]
```diff
@@ -1913,6 +1913,7 @@ static UniValue getblockstats(const JSONRPCRequest& request)
     std::vector<std::pair<CAmount, int64_t>> feerate_array;
     std::vector<int64_t> txsize_array;
 
+    size_t tx_n = 1;
     for (const auto& tx : block.vtx) {
```
Comment (MarcoFalke) [inline src/rpc/blockchain.cpp]: style-nit:

```suggestion
    for (size_t i{0}; i < block.vtx.size(); ++i) {
        const auto& tx = block.vtx.at(i);
        ...// after coinbase check
        const auto& txundo = blockUndo.vtxundo.at(i + 1);
```

**Comment by MarcoFalke** [comment]
@FelixWeis Are you still working on this?

**Comment by jonasschnelli** [comment]
Don't let this go down unmaintained... please pick this up @FelixWeis

File: src/rpc/blockchain.cpp [hunk 8]
```diff
@@ -1879,6 +1879,9 @@ static UniValue getblockstats(const JSONRPCRequest& request)
     }
 
     const CBlock block = GetBlockChecked(pindex);
+    CBlockUndo blockUndo;
+    if (!UndoReadFromDisk(blockUndo, pindex))
```
Comment (MarcoFalke) [inline src/rpc/blockchain.cpp]: style-nit in commit ab4ed66cab35f5339921721cb85b57a1fd378f12:

Could you please add `{}` according to the developer notes and then squash the fixup commit into the first one?

https://github.com/bitcoin/bitcoin/blob/master/CONTRIBUTING.md#squashing-commits

Comment (FelixWeis) [reply src/rpc/blockchain.cpp]: addresses in https://github.com/bitcoin/bitcoin/commit/a565355d8f543cf158138f12130960863736bff3

File: src/rpc/blockchain.cpp [hunk 9]
```diff
@@ -1879,6 +1879,10 @@ static UniValue getblockstats(const JSONRPCRequest& request)
     }
 
     const CBlock block = GetBlockChecked(pindex);
+    CBlockUndo blockUndo;
+    if (!UndoReadFromDisk(blockUndo, pindex)) {
+        throw JSONRPCError(RPC_INTERNAL_ERROR, "Can't read undo data from disk");
+    }
```
Comment (MarcoFalke) [inline src/rpc/blockchain.cpp]:
```suggestion
    }
```

**Review by maflcko — APPROVED** [review]
utACK 293cb0e2ee (only some style comments)

<details><summary>Show signature and timestamp</summary>

Signature:

```
-----BEGIN PGP SIGNED MESSAGE-----
Hash: SHA512

utACK 293cb0e2ee (only some style comments)
-----BEGIN PGP SIGNATURE-----

iQGzBAEBCgAdFiEE+rVPoUahrI9sLGYTzit1aX5ppUgFAlwqrYAACgkQzit1aX5p
pUg0LQv+J5j+966P/KLR0eT22HQJfZGL8PU8dMwq9WEjwhjrLFcccAwMEd4NPGOJ
GaEUbe5/r8V6lJKkXGJcY4Hs874nWKkmrUYVcF9IFGtqDde/75Xf6/cFGQFdMU+Z
9KjKz3Lj1euVeYXCFJ8NNWKhxVwRMSN2+wbqy3/jCnNSjGUc2ob/vrHOXMBA4Peb
gqfRJyUitjmVWB3vM13H0DP4N1gQc3Zd+ILWptZ3KOk147uJZ/YODbTUlbI0JknM
/N8eGu4zbxUR9tMbaG8en3T0otT0/PYEPXy5eBAmjcGC4eHHZLDsbwLaYUkgRCvp
M/WTkBBg5bk+tMbn+hYj4mGp1GCxnm+6D1JuyZsRUOVvu7Zkv87G41Ss+NBq2ozI
SMsIgVJevTh9AeLpD1nPjS2qxWmmbX9WSJGCN/2dRmCi4tTe0dVWsXY8KtJTbVDl
mqL4zWpxGXW0xjimKNDuPw8ONi9BzKKcCT2tn2MA3y4mXbVBeuqkWuMolUw6l3KW
4j4yc522
=/Qza
-----END PGP SIGNATURE-----
```

[Timestamp of file with hash `41e1a1ea262d1060a49ddc699ae7b5453144563ddbb328c375ebb745876632e1  -`](https://opentimestamps.org/info/?004f70656e54696d657374616d7073000050726f6f6600bf89e2e884e89294010841e1a1ea262d1060a49ddc699ae7b5453144563ddbb328c375ebb745876632e1f010433c9e4c06f822c5f97ea1263926c1c908fff01009dd0130fc053338b8ae455a61ad1e7908f12033df27f1202fe933ef00298079c782c9dac5fd7fc1bb42452a08f06b6421a5d008f1045cc73c16f008c8bdb5d5dca40b730083dfe30d2ef90c8e2c2b68747470733a2f2f626f622e6274632e63616c656e6461722e6f70656e74696d657374616d70732e6f7267fff01042ecdd63681fb6a9603f4ce95da76bc508f010c75354fa108cff3d705eb75cc35ebf9608f120957601e841f3e66929b09c2fc27cb4e125f448f182f868c956d34a5e59a23ad208f1045cc73c16f00840ad9b8d8a0732ae0083dfe30d2ef90c8e292868747470733a2f2f66696e6e65792e63616c656e6461722e657465726e69747977616c6c2e636f6dfff010cfc3c6fd2974f60ebd1bcf37dc9b18dc08f1045cc73c16f0087dff30862a81b0fd0083dfe30d2ef90c8e2e2d68747470733a2f2f616c6963652e6274632e63616c656e6461722e6f70656e74696d657374616d70732e6f7267f010f40465109287951f687bc7dd56e8d98208f1045cc73c16f0085418a5fab3e723780083dfe30d2ef90c8e232268747470733a2f2f6274632e63616c656e6461722e636174616c6c6178792e636f6d)

</details>

File: test/functional/rpc_getblockstats.py [hunk 10]
```diff
@@ -55,13 +42,14 @@ def generate_test_data(self, filename):
         mocktime = time.time()
```
Comment (MarcoFalke) [inline test/functional/rpc_getblockstats.py]:
```suggestion
        mocktime = 1525107225
```

nit: could set mocktime to the previous value, so that similar data is generated?

File: test/functional/rpc_getblockstats.py [hunk 11]
```diff
@@ -55,13 +42,14 @@ def generate_test_data(self, filename):
...
-        self.nodes[0].sendtoaddress(address=self.nodes[1].getnewaddress(), amount=10, subtractfeefromamount=True)
+        self.nodes[0].sendtoaddress(address=self.nodes[0].getnewaddress(), amount=10, subtractfeefromamount=True)
         self.nodes[0].generate(1)
         self.sync_all()
 
         self.nodes[0].sendtoaddress(address=self.nodes[0].getnewaddress(), amount=10, subtractfeefromamount=True)
         self.nodes[0].sendtoaddress(address=self.nodes[0].getnewaddress(), amount=10, subtractfeefromamount=False)
-        self.nodes[1].sendtoaddress(address=self.nodes[0].getnewaddress(), amount=1, subtractfeefromamount=True)
+        self.nodes[0].settxfee(amount=0.003)
+        self.nodes[0].sendtoaddress(address=self.nodes[0].getnewaddress(), amount=1, subtractfeefromamount=True)
```
Comment (MarcoFalke) [inline test/functional/rpc_getblockstats.py]: nit could make data gen (slightly more) deterministic:
```suggestion
        self.nodes[0].sendtoaddress(address=self.nodes[0].get_deterministic_priv_key().address, amount=1, subtractfeefromamount=True)
```

**Review by maflcko — COMMENTED** [review]
more test nits

**Comment by MarcoFalke** [comment]
@FelixWeis You don't actually need to rebase. If you address my style-nits, the conflict will solve itself.

**Comment by MarcoFalke** [comment]
Please squash your commits according to https://github.com/bitcoin/bitcoin/blob/master/CONTRIBUTING.md#squashing-commits

**Comment by MarcoFalke** [comment]
re-utACK d20d7567528e216badb8475df298bb3cec008985

Only changes are:
* Make tests more deterministic
* Add and use GetUndoChecked
* Minor stylistic changes

<details><summary>Show signature and timestamp</summary>

Signature:

```
-----BEGIN PGP SIGNED MESSAGE-----
Hash: SHA512

re-utACK d20d7567528e216badb8475df298bb3cec008985

Only changes are:
* Make tests more deterministic
* Add and use GetUndoChecked
* Minor stylistic changes
-----BEGIN PGP SIGNATURE-----

iQGzBAEBCgAdFiEE+rVPoUahrI9sLGYTzit1aX5ppUgFAlwqrYAACgkQzit1aX5p
pUh/4QwAwLZrUZJqLo1ctjMWO2SQ+ut4rxMDYaMNSIGOxw61u8J6OIUPsVfKBJLF
s2Gi0Q7XwCwlttNVDARBCQFH4R8IhVXLaY//BCfVPamVwXCO3s1x7GDtLZZrjs2c
Rng9VOhCSZ7wXZmgoQmYwj77pQOU2k1B6FXkMylzf/pt+W/jaANCuGtC4k4k8x/l
sM2sH75nWs2URra42exUbmDoIfohOHWfAxcA0Ni31AWkeRYRazWGmsvA93qL6PtP
/wN5rMNi10gCI83/N16ZZ+iq7djqRoCjkbkqSNd2D0NJOmquEIN4W3GGQJ1BciGF
nu0gLQ12ShXCVf167E6s4xGPsKu9hsiUNJN9xursiD1RUhGtUQXdpIRjodL+lDj8
CZhHeUj0whAXKkS+YddhPO385rzqChebvdWZL5stRlUi4TFKzJM+rvsjGNaEKgxO
wjiUfi1L6v07zstlKFLc1pfri9Yws5b4MT6fO9+9dQrtUm1IcS4NeS4YrKKaiI4X
B9R6OHC3
=xoBb
-----END PGP SIGNATURE-----
```

[Timestamp of file with hash `5dc2003a827858f9302ac84ff924bdf15ada47e85278dce1eee93d098c8e8abd  -`](https://opentimestamps.org/info/?004f70656e54696d657374616d7073000050726f6f6600bf89e2e884e8929401085dc2003a827858f9302ac84ff924bdf15ada47e85278dce1eee93d098c8e8abdf01001b31b759d70096887582f20c060390608fff0103a16fd4f7c8f24396601d6d3df4b77c308f1045cd5b2aff008685132d3134760540083dfe30d2ef90c8e232268747470733a2f2f6274632e63616c656e6461722e636174616c6c6178792e636f6dfff0107a163f42ff2271de93438f4e185e505108f1045cd5b2aff008b2ee9ad2f7ef444d0083dfe30d2ef90c8e2e2d68747470733a2f2f616c6963652e6274632e63616c656e6461722e6f70656e74696d657374616d70732e6f7267fff01096e0a180f06409fa2b145fd3964ed5f808f1045cd5b2aff0084b4d68869d984add0083dfe30d2ef90c8e2c2b68747470733a2f2f626f622e6274632e63616c656e6461722e6f70656e74696d657374616d70732e6f7267f010ae41a855e90c95af1d865a95e775f1a708f010e234702407890bd5586ea5cfbfa091cb08f1045cd5b2aff008ce6559de452581c00083dfe30d2ef90c8e292868747470733a2f2f66696e6e65792e63616c656e6461722e657465726e69747977616c6c2e636f6d)

</details>
